File: app/test/chat_menu_test.dart
import 'package:comrade/src/util/chat_menu.dart';
import 'package:flutter_test/flutter_test.dart';

/// Pins the conversation ⋮ menu: the comrade and mute rows name the action
/// rather than the state, destructive entries stay last, and nothing is offered
/// twice.
///
/// Mirrors `ChatMenuTest.kt` — a change on one frontend that isn't made on the
/// other shows up as a failing test rather than as silent UI drift.
void main() {
  /// Every combination of the two toggles the menu is built from.
  List<List<ChatMenuAction>> allMenus() => <List<ChatMenuAction>>[
        for (final bool isComrade in <bool>[true, false])
          for (final bool isMuted in <bool>[true, false])
            conversationMenu(isComrade: isComrade, isMuted: isMuted),
      ];

  bool isComradeRow(ChatMenuAction a) =>
      a == ChatMenuAction.addComrade || a == ChatMenuAction.removeComrade;
  bool isMuteRow(ChatMenuAction a) =>
      a == ChatMenuAction.mute || a == ChatMenuAction.unmute;

  group('conversationMenu', () {
    test('a non-comrade is offered the add', () {
      final List<ChatMenuAction> menu = conversationMenu(isComrade: false);
      expect(menu, contains(ChatMenuAction.addComrade));
      expect(menu, isNot(contains(ChatMenuAction.removeComrade)));
    });

    test('a comrade is offered the removal', () {
      final List<ChatMenuAction> menu = conversationMenu(isComrade: true);
      expect(menu, contains(ChatMenuAction.removeComrade));
      expect(menu, isNot(contains(ChatMenuAction.addComrade)));
    });

    test('an unmuted chat is offered mute, and vice versa', () {
      final List<ChatMenuAction> loud =
          conversationMenu(isComrade: false, isMuted: false);
      expect(loud, contains(ChatMenuAction.mute));
      expect(loud, isNot(contains(ChatMenuAction.unmute)));

      final List<ChatMenuAction> quiet =
          conversationMenu(isComrade: false, isMuted: true);
      expect(quiet, contains(ChatMenuAction.unmute));
      expect(quiet, isNot(contains(ChatMenuAction.mute)));
    });

    test('order is stable', () {
      expect(
          conversationMenu(isComrade: false, isMuted: false), <ChatMenuAction>[
        ChatMenuAction.setAlias,
        ChatMenuAction.addComrade,
        ChatMenuAction.mute,
        ChatMenuAction.copyKey,
        ChatMenuAction.encryptionInfo,
        ChatMenuAction.block,
      ]);
    });

    test('toggling comrade or mute swaps one row in place, never reorders', () {
      // Otherwise a tap lands on a different action than the one reached for.
      expect(
          allMenus().map((List<ChatMenuAction> m) => m.length).toSet().length,
          1);
      for (final List<ChatMenuAction> menu in allMenus()) {
        expect(menu.indexWhere(isComradeRow), 1);
        expect(menu.indexWhere(isMuteRow), 2);
      }
    });

    test('no action appears twice', () {
      for (final List<ChatMenuAction> menu in allMenus()) {
        expect(menu.toSet().length, menu.length);
      }
    });

    test('block is the only destructive entry and it is last', () {
      for (final List<ChatMenuAction> menu in allMenus()) {
        expect(menu.last, ChatMenuAction.block);
        expect(
          menu.where((ChatMenuAction a) => a.destructive),
          <ChatMenuAction>[ChatMenuAction.block],
        );
      }
    });

    test('every action is reachable from some menu', () {
      // A menu entry nobody can reach is a dead code path; an action missing
      // from the enum is a non-exhaustive switch at the call site.
      expect(
        allMenus().expand((List<ChatMenuAction> m) => m).toSet(),
        ChatMenuAction.values.toSet(),
      );
    });
  });
}

File: android/app/src/test/java/mullu/comrade/ui/ChatMenuTest.kt
package mullu.comrade.ui

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pins the conversation ⋮ menu: the comrade and mute rows name the action
 * rather than the state, destructive entries stay last, and nothing is offered
 * twice.
 *
 * The Dart port (`app/test/chat_menu_test.dart`) asserts the same order, so a
 * change here that isn't mirrored there shows up as a failing test on the
 * other frontend rather than as silent UI drift.
 */
class ChatMenuTest {

    @Test
    fun nonComradeIsOfferedTheAdd() {
        val menu = conversationMenu(isComrade = false)
        assertTrue(ChatMenuAction.AddComrade in menu)
        assertFalse(ChatMenuAction.RemoveComrade in menu)
    }

    @Test
    fun comradeIsOfferedTheRemoval() {
        val menu = conversationMenu(isComrade = true)
        assertTrue(ChatMenuAction.RemoveComrade in menu)
        assertFalse(ChatMenuAction.AddComrade in menu)
    }

    @Test
    fun unmutedChatIsOfferedMuteAndViceVersa() {
        assertTrue(ChatMenuAction.Mute in conversationMenu(isComrade = false, isMuted = false))
        assertFalse(ChatMenuAction.Unmute in conversationMenu(isComrade = false, isMuted = false))
        assertTrue(ChatMenuAction.Unmute in conversationMenu(isComrade = false, isMuted = true))
        assertFalse(ChatMenuAction.Mute in conversationMenu(isComrade = false, isMuted = true))
    }

    @Test
    fun orderIsStableAndSizeDoesNotChangeWithComradeOrMuteState() {
        assertEquals(
            listOf(
                ChatMenuAction.SetAlias,
                ChatMenuAction.AddComrade,
                ChatMenuAction.Mute,
                ChatMenuAction.CopyKey,
                ChatMenuAction.EncryptionInfo,
                ChatMenuAction.Block,
            ),
            conversationMenu(isComrade = false, isMuted = false),
        )
        // Toggling comrade or mute swaps one row in place — it must not reorder
        // or resize the menu, or a tap lands on a different action than the one
        // the user reached for.
        val sizes = allMenus().map { it.size }.toSet()
        assertEquals(1, sizes.size)
        for (menu in allMenus()) {
            assertEquals(1, menu.indexOfFirst { it.comradeRow })
            assertEquals(2, menu.indexOfFirst { it.muteRow })
        }
    }

    @Test
    fun noActionAppearsTwice() {
        for (menu in allMenus()) {
            assertEquals(menu.size, menu.toSet().size)
        }
    }

    @Test
    fun blockIsTheOnlyDestructiveEntryAndItIsLast() {
        for (menu in allMenus()) {
            assertEquals(ChatMenuAction.Block, menu.last())
            assertEquals(listOf(ChatMenuAction.Block), menu.filter { it.destructive })
        }
    }

    @Test
    fun everyActionIsReachableFromSomeMenu() {
        val offered = allMenus().flatten().toSet()
        // A menu entry nobody can reach is a dead code path; an action missing
        // from the enum is an unhandled `when` branch at the call site.
        assertEquals(ChatMenuAction.entries.toSet(), offered)
    }

    /** Every combination of the two toggles the menu is built from. */
    private fun allMenus(): List<List<ChatMenuAction>> =
        listOf(true, false).flatMap { comrade ->
            listOf(true, false).map { muted -> conversationMenu(isComrade = comrade, isMuted = muted) }
        }

    private val ChatMenuAction.comradeRow: Boolean
        get() = this == ChatMenuAction.AddComrade || this == ChatMenuAction.RemoveComrade

    private val ChatMenuAction.muteRow: Boolean
        get() = this == ChatMenuAction.Mute || this == ChatMenuAction.Unmute
}
